### ISSUE TITLE
ttnn.reshape: ND shard specs abort on rank-reducing reshape (normalized-to-2D input and ND output)

### ISSUE THREAD

**Issue by abdulla-TT** [tenstorrent/tt-metal#56068]
## Summary

`ttnn.reshape` aborts with a shard-rank / missing-shard-spec error in two ND-sharding scenarios:

1. **Normalized ND shard spec on the input** — the input tensor's `MemoryConfig` was built from an `NdShardSpec` that `TensorSpec` normalizes to an equivalent 2D layout (HEIGHT/WIDTH sharded). A rank-reducing reshape aborts even when the requested output is plain interleaved.
2. **ND-sharded output** — the requested output `memory_config` is genuinely ND (`ND_SHARDED`, or carries an `nd_shard_spec` with no 2D `shard_spec`). The reshape aborts before producing the output.

---

## Issue 1 — normalized ND shard spec on the input

```
TT_FATAL @ tt_metal/impl/buffers/buffer_distribution_spec.cpp:335:
tensor_shape.rank() >= shard_shape.rank()
Tensor shape rank (3) can't be less than shard shape rank (4)!
```

### Reproduction

```python
import torch, ttnn

grid = ttnn.CoreRangeSet([ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(1, 0))])
nd = ttnn.NdShardSpec(shard_shape=ttnn.Shape([1, 1, 32, 128]), grid=grid,
                      orientation=ttnn.ShardOrientation.ROW_MAJOR)
mc = ttnn.MemoryConfig(buffer_type=ttnn.BufferType.L1, nd_shard_spec=nd)

t = torch.rand([1, 1, 64, 128], dtype=torch.bfloat16)
x = ttnn.from_torch(t, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT, memory_config=mc)
# x.memory_config().memory_layout reports HEIGHT_SHARDED, but nd_shard_spec (rank 4) is still attached.

ttnn.reshape(x, [1, 1, 128, 64], memory_config=ttnn.DRAM_MEMORY_CONFIG)  # <-- aborts
```

### Root cause

When a `MemoryConfig` is built from a rank-N `NdShardSpec` that is expressible as a 2D layout, `TensorSpec` reports the 2D `memory_layout` (e.g. `HEIGHT_SHARDED`) **but keeps the original higher-rank `nd_shard_spec` attached**.

`reshape` squeezes the input to 3D internally (`reshape_tiled` → `PerformView` → `view_device`) regardless of the caller's target rank. `view_device` (`ttnn/core/tensor/tensor_ops.cpp`) has an ND guard, but it keys on `memory_layout() == ND_SHARDED`, so these normalized configs bypass it into the 2D branch, which carries the input's memory config (with its rank-4 `nd_shard_spec`) onto the now rank-3 tensor. `BufferDistributionSpec` then enforces `shard rank <= tensor rank` and aborts (4 > 3).

The abort originates from the **input** rank reduction, not from output handling; the output in the repro is interleaved and never ND.

---

## Issue 2 — ND-sharded output

```
TT_FATAL @ ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
memory_config.shard_spec().has_value()
```

### Reproduction

```python
import torch, ttnn

grid = ttnn.CoreRangeSet([ttnn.CoreRange(ttnn.CoreCoord(0, 0), ttnn.CoreCoord(1, 0))])
nd = ttnn.NdShardSpec(shard_shape=ttnn.Shape([1, 1, 64, 64]), grid=grid,
                      orientation=ttnn.ShardOrientation.ROW_MAJOR)
out_mc = ttnn.MemoryConfig(buffer_type=ttnn.BufferType.L1, nd_shard_spec=nd)

t = torch.rand([1, 1, 64, 128], dtype=torch.bfloat16)
x = ttnn.from_torch(t, dtype=ttnn.bfloat16, device=device, layout=ttnn.TILE_LAYOUT,
                    memory_config=ttnn.DRAM_MEMORY_CONFIG)

ttnn.reshape(x, [1, 1, 128, 64], memory_config=out_mc)  # <-- aborts
```

### Root cause

The internal sharded reshape paths derive the output shard spec from a 2D `shard_spec` (`recompute_shard_spec_for_output`) and realize it via `interleaved_to_sharded`. Neither handles an ND shard spec, so an output `memory_config` that is genuinely ND (`ND_SHARDED`, or an `nd_shard_spec` with no 2D `shard_spec`) hits the `shard_spec().has_value()` assertion and aborts. This applies whether the ND spec normalizes to a 2D layout on the output tensor or stays genuinely ND (e.g. a batch split).

**Comment by abdulla-TT** [comment]
## Update: all three reshape sharding gaps from the audit now fixed

Scope expanded from the original single ND-input bug to cover all three reshape entries flagged in the audit. All verified on wormhole_b0 (PCC 1.0 vs torch golden) for TILE and ROW_MAJOR.

**1. ND-sharded input that normalizes to a 2D layout** (original bug)
- Root cause: `view_device` (`ttnn/core/tensor/tensor_ops.cpp`) carried a retained higher-rank `nd_shard_spec` onto a rank-reduced tensor.
- Fix: in the 2D branch, drop the `nd_shard_spec` when its rank exceeds the new padded shape's rank, letting the equivalent 2D `shard_spec` describe the buffer.

**2. DRAM-sharded ROW_MAJOR input** (was "RM fails" in §3a)
- Root cause: the RM reshape path un-shards the input via `sharded_to_interleaved`, which is L1-only, so a DRAM-sharded input aborted with `"Input tensor must be in L1"`.
- Fix: `perform_reshape_on_2D_RM` now un-shards via `ttnn::to_memory_config`, which routes L1-sharded through the same `sharded_to_interleaved` and DRAM-sharded through `prim::copy`. No behavior change for L1.

**3. ND-sharded output** (§3b)
- Root cause: the internal sharded paths derive the output shard spec from a 2D `shard_spec` (`recompute_shard_spec_for_output`) and realize it via `interleaved_to_sharded`, neither of which handles an ND shard spec; ND output configs aborted at `reshape_view/reshape.cpp:109`.
- Fix: top-level `ttnn::reshape` intercepts ND-sharded output configs (`memory_layout == ND_SHARDED`, or an `nd_shard_spec` with no 2D `shard_spec`), reshapes to an interleaved intermediate, then lays out the ND output via `to_memory_config`. Layout-only sharded outputs (no shard_spec, non-ND) are left on the existing auto-derive path.

### Verification
- 10 new regression tests in `test_universal_input_tm_reshape.py` (ND input, DRAM-sharded RM input, ND output incl. a genuinely-ND batch-split case), TILE + RM.
- No regressions: reshape universal-input suite (348), base reshape suites (335 + skips/xfails), `test_tensor_nd_sharding.py` (428).

### Correction to the audit
The audit attributed the §3b ND-**output** failure to `buffer_distribution_spec.cpp:335`. That line is actually the cause of the ND-**input** case (#1); the ND-output case fails earlier, at `reshape_view/reshape.cpp:109`.
